File: algorithms/algorithms-java/src/main/java/com/lanceolata/coding/algorithms/java/coding_interviews_2/Question_0021_01_ReorderArray.java
package com.lanceolata.coding.algorithms.java.coding_interviews_2;

public class Question_0021_01_ReorderArray {
    public void reOrderArray(int [] array) {
        if (array == null) {
            return;
        }
        int l = 0, r = array.length - 1;
        while (l < r) {
            while (l < r && isOdd(array[l])) {
                l++;
            }
            while (l < r && !isOdd(array[r])) {
                r--;
            }
            if (l < r) {
                int tmp = array[l];
                array[l] = array[r];
                array[r] = tmp;
            }
        }
    }

    private boolean isOdd(int n) {
        return (n & 1) != 0;
    }
}

File: algorithms/algorithms-cplusplus/coding_interviews_2/Question_0021_01_ReorderArray.cc
#include <vector>

using namespace std;

class Solution {
 public:
  void reOrderArray(vector<int> &array) {
    if (array.size() <2) {
      return;
    }
    int l = 0, r = array.size() - 1;
    while (l < r) {
      while (l < r && isOdd(array[l])) {
        l++;
      }
      while (l < r && !isOdd(array[r])) {
        r--;
      }
      if (l < r) {
        swap(array[l], array[r]);
      }
    }
  }

  bool isOdd(int n) {
    return (n & 1) != 0;
  }
};
